[ryanongwx] ip

ryanongwx opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:
- [ ] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

``public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}``

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ryanongwx committed:
Level-0 Rename, Greet, Exit

ryanongwx committed:
Level-1 Echo

ryanongwx committed:
Level-2 Add, List

ryanongwx committed:
Level-3 Mark as Done

ryanongwx committed:
Level-4 ToDos, Events, Deadlines

ryanongwx committed:
A-TextUiTesting

ryanongwx committed:
Level-5 Handle Errors

ryanongwx committed:
Level-6 Delete

ryanongwx committed:
Level-7 Save

ryanongwx committed:
Level-8 Dates and Times

ryanongwx committed:
A-Packages

ryanongwx committed:
Merge branch 'master' into add-gradle-support

Merge

ryanongwx committed:
A-Gradle

ryanongwx committed:
A-JUnit

ryanongwx committed:
A-JavaDoc

ryanongwx committed:
Level-9 Find

ryanongwx committed:
Level 10. GUI
Add a GUI to the chatbot using JavaFX technology

In the current situation, the chatbot lacks a graphical user interface (GUI), making it less user-friendly and accessible.

To address this issue, this commit introduces a GUI to the chatbot. We have chosen to implement the GUI using JavaFX technology due to its robustness and cross-platform compatibility.

The GUI will enhance the user experience by providing an intuitive way to interact with the chatbot. Users will now have a visually appealing interface to input queries and receive responses, making the application more user-friendly.

This approach is chosen because JavaFX offers a wide range of features for building interactive and responsive interfaces. It allows us to create a seamless integration between the chatbot's functionality and the graphical interface, ensuring a cohesive user experience.

Additionally, JavaFX is well-documented and has a strong community, which will facilitate future maintenance and updates.

In summary, this commit adds a JavaFX-based GUI to the chatbot, improving its usability and overall user experience.

- Tested on Windows, macOS, and Linux.
- GUI design follows best practices for user interface usability.
- Dependencies updated to include JavaFX libraries.

mfjkri commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    public void parseDateTime(String dateTime) {
```
Consider writing a descriptive header comments for all public methods.

mfjkri commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,121 @@
+    public static String parseAndAddTask(String userInput, TaskList taskList) {
+        String[] words = userInput.split(" ", 2);
+
+        switch (words[0].toLowerCase()) {
```
Consider removing the indentation for the case clauses to adhere to coding standard.

mfjkri commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
Consider writing a descriptive header comments for all public classes.

mfjkri commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,148 @@
+        userInput = new TextField();
+        sendButton = new Button("Send");
+
+        AnchorPane mainLayout = new AnchorPane();
```
Is there a better place to put this logic? 

Perhaps another class such as MainWindow?

mfjkri reviewed (commented):
Good code readability and comments!

TeeRenJing commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+
+public class Deadline extends Task {
+    protected LocalDate byDate;
+    protected LocalTime byTime;
```
Any reason why the fields are protected instead of private?

TeeRenJing commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+        this.byDateTimeString = byDateAndTime;
+    }
+
+    public boolean isValidDate(String dateTimeString) {
```
I like how boolean methods are named to sound like booleans 👍

TeeRenJing commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,69 @@
+package duke;
+
+import java.io.*;
```
Should the imported classes be listed explicitly?

TeeRenJing commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,37 @@
+package duke;
```
I like how classes are put in packages! 🥇

TeeRenJing reviewed (commented):
Looks good to me 👍 
Just a few nits to fix.

ryanongwx committed:
***A-Assertions***
Use Assertions

In the current situation, our codebase lacks a systematic way of documenting and validating important assumptions at critical points in the code.

To address this issue, this commit introduces the use of assertions as a means of documenting and validating these assumptions. Assertions serve as checkpoints within the code, helping ensure that specific conditions or expectations are met during program execution.

The inclusion of assertions enhances the code's clarity by explicitly stating the expectations and assumptions made by the developer. This not only aids in understanding the code but also helps catch potential issues early in the development process.

Here's why we're using assertions:

1. **Documentation**: Assertions provide in-code documentation for crucial assumptions, making it easier for developers to understand the code's requirements and constraints.

2. **Debugging**: When an assertion fails, it provides valuable information about what went wrong and where, simplifying the debugging process.

3. **Safety Nets**: Assertions act as safety nets, catching unexpected conditions or bugs during testing and development, reducing the chances of releasing faulty code.

4. **Maintenance**: Over time, assertions facilitate code maintenance by helping developers quickly identify the impact of changes on existing assumptions.

To implement assertions, we're using the built-in assert feature of the programming language (not JUnit assertions). This allows us to assert conditions directly in the code without introducing external dependencies. Assertions should be used judiciously, primarily for scenarios where the failure of an assumption would lead to severe issues or incorrect behavior.

Here's an example of how assertions are used:

```java
public int divide(int dividend, int divisor) {
    assert divisor != 0 : "Divisor should not be zero!";
    // Rest of the division logic here
}

ryanongwx committed:
A-CodeQuality

ryanongwx committed:
***B-Reminders***
Reminders for tasks

In the current situation, our task management system lacks a mechanism for users to receive reminders about upcoming deadlines, which is essential for helping users stay organized and meet their commitments.

To address this issue, this commit introduces a new feature that provides reminders for tasks. Specifically, it adds an "Upcoming Deadlines" section to the top of the todo list. This section calculates and displays the number of days left until each task's deadline, serving as a convenient and visible reminder for users.

Here's why we've added this feature:

1. **Improved Task Management**: Reminders help users keep track of their deadline dates, reducing the risk of missing important deadelines.

2. **Enhanced User Experience**: The "Upcoming Deadlines" section adds value to our application by making it more user-friendly and efficient.

3. **Increased Productivity**: Reminders promote better time management and productivity by ensuring that users are aware of their upcoming deadlines.

4. **Efficiency**: Displaying the number of days left to each deadline provides users with a clear understanding of their task priorities.

The implementation of this feature includes the following:

- The "Upcoming Deadlines" section is added to the top of the todo list interface.

- A calculation is performed to determine the number of days remaining until each task's deadline.

This approach ensures that our reminder system is both informative and user-configurable, catering to different user needs and preferences.

In summary, this commit adds a valuable "Upcoming Deadlines" feature to our task management system, providing users with reminders for their tasks' deadlines and enhancing their overall experience.

ryanongwx committed:
Minor UI fixes

ryanongwx committed:
Merge pull request #2 from ryanongwx/branch-UIFixes

Minor UI fixes

ryanongwx committed:
A-UserGuide

ryanongwx committed:
Merge pull request #3 from ryanongwx/branch-A-UserGuide

A-UserGuide

ryanongwx committed:
final changes

ryanongwx committed:
Merge branch 'branch-A-UserGuide'
